src/java.base/share/classes/java/net/Socket.java

```diff
@@ -497,13 +497,10 @@ private Socket(SocketAddress address, SocketAddress localAddr, boolean stream)
             if (localAddr != null)
                 bind(localAddr);
             connect(address);
-        } catch (IOException | IllegalArgumentException | SecurityException e) {
-            try {
-                close();
-            } catch (IOException ce) {
-                e.addSuppressed(ce);
-            }
-            throw e;
+        } catch (Throwable throwable) {
+            // `connect()` already closes the socket on failures, yet `bind()` doesn't, and hence:
+            closeQuietly(throwable);
+            throw throwable;
```

Comment:
Would you mind adding `assert address instanceof InetSocketAddress isa && !isa.isUnresolved();` to the start of this (private) constructor as that will ensure we have the right pre-conditions and avoid further discussion about other socket address types.

Reply:
If you take on board the other feedback then you can reduce the scope of the try-catch to just the bind, e.g.
```
        if (localAddr != null) {
            try {
                bind(localAddr);
            } catch (Throwable ex) {
                try {
                    close();
                } catch (Throwable ce) { ex.addSuppressed(ce); }
                throw ex;
            }
        }
        connect(address);   // closes Socket if it throws
```

Reply:
Fixed in 424fc8ccd3a495a49f859f0cfc563cf463b3efeb.

Reply:
I don't mind if we have a broader try-finally but it would be a bug if connect were to throw IAE. All the validation should be done before delegating to the SocketImpl.

Reply:
Extended the try-finally to cover both `bind()` and `connect()` exceptions in ce568022f9a0fc9c0e9727911e3856048e1a054c.

Reply:
Mark Sheppard spotted that too, see exchange in JBS issue. You can drop the !isUnresolved from the assert or eagerly throw UHE if attempting to creating a Socket with an unresolved address, either is fine but probably simple to adjust the assert. The downside is that if someone creates a Socket with an unresolved address then it will create a SocketImpl and attempt to connect before it detects the bad input. It's a corner case so not a big deal.

Reply:
Yes, I was indeed pinged by @msheppar. (Sorry, I missed the JBS conversation.) Removed the `&& !isa.isUnresolved()` from the assertion in 6b13675de44fd24766d355929f9b695a48cacc71.